src/google/protobuf/compiler/csharp/csharp_helpers.cc

```diff
@@ -381,13 +381,22 @@ std::string GetFieldConstantName(const FieldDescriptor* field) {
 std::string GetPropertyName(const FieldDescriptor* descriptor) {
   // TODO(jtattermusch): consider introducing csharp_property_name field option
   std::string property_name = UnderscoresToPascalCase(GetFieldName(descriptor));
-  // Avoid either our own type name or reserved names. Note that not all names
-  // are reserved - a field called to_string, write_to etc would still cause a problem.
+  // Avoid either our own type name or reserved names.
   // There are various ways of ending up with naming collisions, but we try to avoid obvious
-  // ones.
+  // ones. In particular, we avoid the names of all the members we generate or inherit.
   if (property_name == descriptor->containing_type()->name()
       || property_name == "Types"
-      || property_name == "Descriptor") {
+      || property_name == "Descriptor"
+      || property_name == "Equals"
+      || property_name == "ToString"
+      || property_name == "GetHashCode"
+      || property_name == "WriteTo"
+      || property_name == "Clone"
+      || property_name == "CalculateSize"
+      || property_name == "Descriptor"
+      || property_name == "MergeFrom"
+      || property_name == "OnConstruction"
+      || property_name == "Parser") {
```

Comment:
perhaps we have gotten to the point where a hash set or makes sense?

Reply:
I would have done that, but my C++ skills are incredibly poor :) If you have an example elsewhere in the codebase that I could refer to, that would be really handy.

Reply:
```
std::unordered_set<std::string> reserved_names = {
  descriptor->containing_type()->name(),
  "Types",
  ...
};
```

Reply:
```
if (reserved_names->find(name) != reserved_names->end()) {
  ....
}
```

Reply:
the invocation for a static one is:

```
static const auto& reserved_names = *new std::unordered_set<std::string>({"hi", "there"});
```

Reply:
Many thanks. I'll give each a try tomorrow, and ping the PR when I've come up with something. Thanks for putting up with my C++ ignorance :)

Reply:
Sorted now, thanks.